### PR TITLE
[FSDP] Break up `_post_backward_hook` into smaller funcs

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107784
* #106080
* __->__ #106068

The post-backward hook has some complexity due to the different paths: {no communication hook, communication hook} x {`NO_SHARD`, `FULL_SHARD`/`SHARD_GRAD_OP`, `HYBRID_SHARD`/`_HYBRID_SHARD_ZERO2`} plus some options like CPU offloading and `use_orig_params=True` (requiring using sharded gradient views).

The PR following this one that adds async all-reduce for HSDP further complicates this since the bottom-half after all-reduce must still be run in the separate all-reduce stream, making it more unwieldy to unify with the existing bottom-half.

Nonetheless, this PR breaks up the post-backward hook into smaller logical functions to hopefully help readability.

Differential Revision: [D47852461](https://our.internmc.facebook.com/intern/diff/D47852461)

### CODE REVIEW

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47852461).

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -951,6 +833,154 @@ def _should_free_in_backward(
     )
```
Comment (ezyang) [inline torch/distributed/fsdp/_runtime_utils.py:835]: Are you perhaps willing to add small top level comments for each of the new functions? :)

**Review by ezyang — APPROVED** [review]
Verified code motion only

File: torch/distributed/fsdp/_runtime_utils.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:835]: I added the comments in the subsequent PR since the async HSDP refactors these a bit more now.

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**